scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,296 @@
+        if (scanData.entities && scanData.entities.length > 0) {
+            result.isEmpty = false;
+            result.entities = scanData.entities;
+        }
+
+        // If the response contains an "after" attribute inside "cursors" (cursors.after), then, there are still some External Contacts (data) to retrieve
+        // When the response does not contain an "after" attribute inside "cursors", it means it is the last page of scan
+        if (scanData.hasOwnProperty("cursors") && scanData.cursors.hasOwnProperty("after")) {
+            result.isFinished = false;
+            result.nextCursor = scanData.cursors.after;
+        }
+
+        // Reset of some global vars - counters for retry
+        consecutiveRateLimitRetryAttempts = 0;
+
+        return result;
+    }
+    catch (error) {
+        console.log('External Contacts Scan error - Error:', JSON.stringify(error, null, 2));
+
+        if (error.status && error.status === 429) {
+            // API Rate Limit - Max number of API requests per minute reached
+            // Exit Application if too many consecutive retry failures
+            if (consecutiveRateLimitRetryAttempts >= maxRateLimitRetryAttempts) {
+                result.errorMsg = "Max consecutive API Rate Limit. Exiting...";
+                result.error = true;
+                return result;
+            } else {
+                // Extract retry-after header value (seconds) from the 429 error
+                let retryAfter = 0;
+                if (error.headers && error.headers.hasOwnProperty("retry-after")) {
+                    retryAfter = parseInt(error.headers["retry-after"]);
+                } else {
+                    result.errorMsg = "Failed to retrieve Retry-After. Exiting...";
+                    result.error = true;
+                    return result;
+                }
+                // Add 1 second (for "safety")
+                let retryIn = (retryAfter + 1) * 1000;
+                // Pause execution
+                await new Promise((resolve, reject) => setTimeout(resolve, retryIn));
+
+                consecutiveRateLimitRetryAttempts++;
+
+                // Try to request the same data again (current value of cursor)
+                let retryScanData = await scanContacts(cursor);
+                return retryScanData;
+            }
+        }
+        else {
+            // 400 Bad Request - ex: cursor invalid/expired
+            // 401 Unauthorized - ex: Invalid token, Token expired
+            // and other errors
+            result.errorMsg = "Unexpected Error. Exiting...";
+            result.error = true;
+            return result;
+        }
+    }
+}
+
+//#endregion
+
+/*
+------------------------------------------------------------------------------
+* Main
+------------------------------------------------------------------------------
+*/
+
+(async () => {
+
+    if (maxRecordsPerPage < 10 || maxRecordsPerPage > 200) {
```

Comment:
This method is really too big.  I like to break my "main" method into a series of function calls that simply step through each step.  I would break this funtion into smaller pieces and then walk through each function.